### PR TITLE
fix unhandled runtime.lastError for tcp socket closing

### LINKED ISSUE #90: TCP socket close behaviour has changed
I'm running Chrome 53.0.2785.116 on Linux.

Consider a simple programme which:
 * creates a client socket
 * connects to some server
 * immediately closes that socket

Two issues:
 1. unchecked `runtime.lastError` in the javascript console (it comes from `dispatchDisconnect`: https://github.com/freedomjs/freedom-for-chrome/blob/master/providers/core.tcpsocket.js#L419)
 1. `dispatchDisconnect` dispatches the wrong `errcode`: `CONNECTION_CLOSED`, when it should be `SUCCESS` (the problem is Chrome is returning -100 rather than 0 to `handleReadError`: https://github.com/freedomjs/freedom-for-chrome/blob/master/providers/core.tcpsocket.js#L519)

### PR DESCRIPTION
**Pull request by soycode**
Fix #90 (depends on/tested by https://github.com/freedomjs/freedom/pull/327)

### CODE REVIEW

**Comment by soycode** [comment]
Noting here for my edification - best way to fully test this is by pulling it into uProxy and doing full integration/simple socks. Also, it may be worth investigating the `dispatchDisconnect` method as related to this, in order to truly catch the runtime error and not just relabel it.

File: providers/core.tcpsocket.js [hunk 1]
```diff
@@ -416,7 +416,12 @@ Socket_chrome.prototype.dispatchDisconnect = function (code) {
   if (this.hasId()) {
     // Every socket must be explicitly closed, even if it has already been
     // disconnected, to avoid a memory leak.
-    this.namespace.close(this.id, function() {});
+    this.namespace.close(this.id, function() {
+      if (chrome.runtime.lastError) {
+        // https://github.com/freedomjs/freedom-for-chrome/issues/90
+        console.warn("Ignoring runtime.lastError, socket is closed");
```
Comment (trevj) [inline providers/core.tcpsocket.js]: If it's harmless, it shouldn't be logged as `warn` :-)

File: providers/core.tcpsocket.js [hunk 2]
```diff
@@ -517,6 +522,11 @@ Socket_chrome.handleReadData = function (readInfo) {
  * @static
  */
 Socket_chrome.handleReadError = function (readInfo) {
+  if (readInfo.resultCode === -100) {
```
Comment (trevj) [inline providers/core.tcpsocket.js]: This doesn't fix the uProxy integration tests.

Digging down a little, it seems like `handleReadError` is being called with -100 whenever a socket terminates - maybe even before the `close` callback.

I created a commit to help debug this:
https://github.com/uProxy/uproxy/compare/trevj-tcp-socket-close-debugging?expand=1

Can you patch this in and experiment a little?

**Comment by soycode** [comment]
I believe I figured it out - we should only override the error message on the side where we're purposely closing the socket, and not on the remote/server side. With this change, both the new freedom.js test and the existing uProxy TCP tests pass, for me at least. Let me know if it looks good to you, thanks!

**Comment by trevj** [comment]
Yay, this seems to fix the integration tests!

**Comment by trevj** [comment]
:+1:

**Pull request merged**